### ISSUE TITLE
TypeError: __details_cb() missing 5 required positional arguments

### ISSUE THREAD

**Issue by gfxmonk** [0install/0install#32]
After upgrading to fedora 21, looks like the PackageKit dbus or python API has changed a bit. When I run the python version (http://0install.net/2007/interfaces/ZeroInstall.xml, v2.3.4), I get the following error printed to the console (but the process continues):

    ERROR:dbus.connection:Exception in handler for D-Bus signal:
    Traceback (most recent call last):
      File "/usr/lib64/python3.4/site-packages/dbus/connection.py", line 230, in maybe_handle_message
        self._handler(*args, **kwargs)
    TypeError: __details_cb() missing 5 required positional arguments: 'licence', 'group', 'detail', 'url', and 'size'

[Here is the output from `dbus-monitor`](https://gist.github.com/gfxmonk/25305a0b48fd1ec0d19c) while I ran:

    0install run --command=0install http://0install.net/2007/interfaces/ZeroInstall.xml select -rc http://gfxmonk.net/dist/0install/git.xml

PackageKit resolution seems to work fine with the ocaml version, so I guess it's an API change on python's side?

**Comment by talex5** [comment]
Guess so. Try changing `__details_cb` in `packagekit.py`. Maybe it takes a single dict argument now or something?

**Comment by gfxmonk** [comment]
Yep, looks like it's now being called with a single argument. `kwargs` is empty, while `args` is:

    args=[dbus.Dictionary(
    {
      dbus.String(u'package-id'): dbus.String(u'git;2.1.0-2.fc21;x86_64;fedora', variant_level=1),
      dbus.String(u'description'): dbus.String(u'Git is a fast, scalable, distributed revision control system with an unusually rich command set that provides both high-level operations and full access to internals.  The git rpm installs the core tools with minimal dependencies.  To install all git packages, including tools for integrating with other SCMs, install the git-all meta-package.', variant_level=1),
      dbus.String(u'license'): dbus.String(u'GPLv2', variant_level=1),
      dbus.String(u'url'): dbus.String(u'http://git-scm.com/', variant_level=1),
      dbus.String(u'summary'): dbus.String(u'Fast Version Control System', variant_level=1),
      dbus.String(u'size'): dbus.UInt64(5108512L, variant_level=1)
    }, signature=dbus.Signature('sv'))]


So perhaps a compatible implementation of __details_cb might be:

	def __details_cb(self, *args):
		if len(args) == 1:
			# recent PackageKit versions pass a single dictionary, rather than individual args
			id      = args[0]['package-id']
			licence = args[0]['license']
			group   = args[0]['group']
			detail  = args[0]['detail']
			url     = args[0]['url']
			size    = args[0]['size']
		else:
			id, licence, group, detail, url, size = args
                        # XXX this assumes the old PackageKit API uses only `args`, not `kwargs`

		details = {'licence': str(licence),
				   'group': str(group),
				   'detail': str(detail),
				   'url': str(url),
				   'size': int(size)}
		_logger_pk.debug(_('Details: %s %r'), id, details)
		self.details[id] = details

..except that the `group` and `detail` keys are not present. The docs at http://www.freedesktop.org/software/PackageKit/gtk-doc/Transaction.html#Transaction::Details say that `group` and `detail` should be present in the dict, but they also say that `package-id` should be `package_id`.

**Comment by talex5** [comment]
As far as I can see, nothing uses `group` or `detail` anyway. If it's working for you, please go ahead and submit a pull request.
